contracts/icp/context-config/src/types.rs

```diff
@@ -0,0 +1,346 @@
+        self.0.as_bytes()
+    }
+
+    fn from_bytes<F>(f: F) -> repr::Result<Self, Self::Error>
+    where
+        F: FnOnce(&mut Self::DecodeBytes) -> Bs58Result<usize>,
+    {
+        Identity::from_bytes(f).map(Self)
+    }
+}
+
+#[derive(CandidType, Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
+pub struct ICApplication {
+    pub id: ICApplicationId,
+    pub blob: ICBlobId,
+    pub size: u64,
+    pub source: String,
+    pub metadata: Vec<u8>,
+}
+
+impl From<Application<'_>> for ICApplication {
+    fn from(value: Application) -> Self {
+        ICApplication {
+            id: value.id.rt().expect("infallible conversion"),
+            blob: value.blob.rt().expect("infallible conversion"),
+            size: value.size,
+            source: value.source.0.into_owned(),
+            metadata: value.metadata.0.into_inner().into_owned(),
+        }
+    }
+}
+
+impl<'a> From<ICApplication> for Application<'a> {
+    fn from(value: ICApplication) -> Self {
+        Application::new(
+            value.id.rt().expect("infallible conversion"),
+            value.blob.rt().expect("infallible conversion"),
+            value.size,
+            ApplicationSource(Cow::Owned(value.source)),
+            ApplicationMetadata(Repr::new(Cow::Owned(value.metadata))),
+        )
+    }
+}
+
+#[derive(CandidType, Serialize, Deserialize, Debug, Clone)]
+pub struct ContextRequest {
+    pub context_id: ICContextId,
+    pub kind: ContextRequestKind,
+}
+
+#[derive(CandidType, Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+pub enum ICCapability {
+    ManageApplication,
+    ManageMembers,
+    Proxy,
+}
+
+#[derive(CandidType, Serialize, Deserialize, Debug, Clone)]
+pub enum ContextRequestKind {
+    Add {
+        author_id: ICContextIdentity,
+        application: ICApplication,
+    },
+    UpdateApplication {
+        application: ICApplication,
+    },
+    AddMembers {
+        members: Vec<ICContextIdentity>,
+    },
+    RemoveMembers {
+        members: Vec<ICContextIdentity>,
+    },
+    Grant {
+        capabilities: Vec<(ICContextIdentity, ICCapability)>,
+    },
+    Revoke {
+        capabilities: Vec<(ICContextIdentity, ICCapability)>,
+    },
+    UpdateProxyContract,
+}
+
+#[derive(CandidType, Serialize, Deserialize, Debug, Clone)]
+pub enum RequestKind {
+    Context(ContextRequest),
+}
+
+#[derive(CandidType, Serialize, Deserialize, Debug, Clone)]
+pub struct Request {
+    pub kind: RequestKind,
+    pub signer_id: ICSignerId,
+    pub timestamp_ms: u64,
+}
+
+impl Request {
+    pub fn new(signer_id: ICSignerId, kind: RequestKind) -> Self {
+        Self {
+            signer_id,
+            kind,
+            timestamp_ms: 0, // Default timestamp for tests
+        }
+    }
+
+    #[cfg(test)]
+    pub fn new_with_time(signer_id: ICSignerId, kind: RequestKind, timestamp_ms: u64) -> Self {
+        Self {
+            signer_id,
+            kind,
+            timestamp_ms,
+        }
+    }
```

Comment:
why do we have 2 of these?

Reply:
Ideally we shouldn't have special behaviour for tests, so it more precisely reflects production behaviour

Reply:
I had some issues with time generation but this is legacy. 
Removed it from code